src/simulation/dynamics/constraintEffector/_UnitTest/test_ConstraintDynamicEffectorUnit.py

```diff
@@ -0,0 +1,331 @@
+    scObject1.hub.omega_BN_BInit = omega_B1N_B1
+    scObject2.hub.r_CN_NInit = r_B2N_N
+    scObject2.hub.v_CN_NInit = rDot_B2N_N
+    scObject2.hub.sigma_BNInit = sigma_B2N
+    scObject2.hub.omega_BN_BInit = omega_B2N_B2
+
+    # Create the constraint effector module
+    constraintEffector = constraintDynamicEffector.ConstraintDynamicEffector()
+    # Set up the constraint effector
+    constraintEffector.r_P1B1_B1 = r_P1B1_B1
+    constraintEffector.r_P2B2_B2 = r_P2B2_B2
+    constraintEffector.r_P2P1_B1Init = r_P2P1_B1Init
+    constraintEffector.alpha = 1e3
+    constraintEffector.beta = constraintEffector.alpha
+    constraintEffector.ModelTag = "constraintEffector"
+    # Add constraints to both spacecraft
+    scObject1.addDynamicEffector(constraintEffector)
+    scObject2.addDynamicEffector(constraintEffector)
+
+    # Add test module to runtime call list
+    unitTestSim.AddModelToTask(unitTaskName, scObject1)
+    unitTestSim.AddModelToTask(unitTaskName, scObject2)
+    unitTestSim.AddModelToTask(unitTaskName, constraintEffector)
+
+    # Record the spacecraft state message
+    datLog1 = scObject1.scStateOutMsg.recorder()
+    datLog2 = scObject2.scStateOutMsg.recorder()
+    unitTestSim.AddModelToTask(unitTaskName, datLog1)
+    unitTestSim.AddModelToTask(unitTaskName, datLog2)
+
+    # Log energy and momentum variables
+    conservationData1 = scObject1.logger(["totOrbAngMomPntN_N", "totRotAngMomPntC_N", "totOrbEnergy", "totRotEnergy"])
+    conservationData2 = scObject2.logger(["totOrbAngMomPntN_N", "totRotAngMomPntC_N", "totOrbEnergy", "totRotEnergy"])
+    unitTestSim.AddModelToTask(unitTaskName,conservationData1)
+    unitTestSim.AddModelToTask(unitTaskName,conservationData2)
+
+    # Initialize the simulation
+    unitTestSim.InitializeSimulation()
+
+    # Setup and run the simulation
+    stopTime = 1
+    unitTestSim.ConfigureStopTime(macros.sec2nano(stopTime))
+    unitTestSim.ExecuteSimulation()
+
+    # collect the recorded spacecraft states
+    constraintTimeData = datLog1.times() * macros.NANO2SEC
+    r_B1N_N_hist = datLog1.r_BN_N
+    rdot_B1N_N_hist = datLog1.v_BN_N
+    sigma_B1N_hist = datLog1.sigma_BN
+    omega_B1N_B1_hist = datLog1.omega_BN_B
+    r_B2N_N_hist = datLog2.r_BN_N
+    rdot_B2N_N_hist = datLog2.v_BN_N
+    sigma_B2N_hist = datLog2.sigma_BN
+    omega_B2N_B2_hist = datLog2.omega_BN_B
+
+    # collect the logged conservation variables
+    conservationTimeData = conservationData1.times() * macros.NANO2SEC
+    orbEnergy1 = conservationData1.totOrbEnergy
+    orbAngMom1_N = conservationData1.totOrbAngMomPntN_N
+    rotAngMom1PntC1_N = conservationData1.totRotAngMomPntC_N
+    rotEnergy1 = conservationData1.totRotEnergy
+    orbEnergy2 = conservationData2.totOrbEnergy
+    orbAngMom2_N = conservationData2.totOrbAngMomPntN_N
+    rotAngMom2PntC2_N = conservationData2.totRotAngMomPntC_N
+    rotEnergy2 = conservationData2.totRotEnergy
+
+    # Compute constraint violations
```

Comment:
I wonder if having this as a separate function would be better. This code is quite complex.